nexus/test-utils/src/background.rs

```diff
@@ -71,19 +53,40 @@ pub async fn wait_background_task(
 }
 
 /// Given the name of a background task, activate it, then wait for it to
-/// complete. Return the last polled `BackgroundTask` object.
+/// complete. Return the `BackgroundTask` object from this invocation.
 pub async fn activate_background_task(
     internal_client: &ClientTestContext,
     task_name: &str,
 ) -> BackgroundTask {
-    let task = NexusRequest::object_get(
-        internal_client,
-        &format!("/bgtasks/view/{task_name}"),
-    )
-    .execute_and_parse_unwrap::<BackgroundTask>()
-    .await;
+    // If it is running, wait for an existing task to complete - this function
+    // has to wait for _this_ activation to finish.
+    //
+    // If it has never run, this function will return straight away.
+    let previous_task = wait_for_condition(
+        || async {
+            let task = NexusRequest::object_get(
+                internal_client,
+                &format!("/bgtasks/view/{task_name}"),
+            )
+            .execute_and_parse_unwrap::<BackgroundTask>()
+            .await;
+
+            if matches!(task.current, CurrentStatus::Idle) {
+                return Ok(task);
+            }
 
-    let last_activate = most_recent_activate_time(&task);
+            info!(
+                internal_client.client_log,
+                "waiting for {task_name} to go idle",
+            );
+
+            Err(CondCheckError::<()>::NotYet)
+        },
+        &Duration::from_millis(50),
+        &Duration::from_secs(10),
+    )
+    .await
+    .expect("task went to idle");
```

Comment:
Nit - this text isn't correct, right? We don't have a way of returning a permanent error from this function

Reply:
Ah, I always write these type of messages in the positive, but that isn't right here: "task went to idle: timed out" (for example) isn't the right message for this. Changed in d6a60bec8